README
WebCalender 0.9.29 - 05 Jun 01
http://webcalendar.sourceforge.net
Craig Knudsen, cknudsen@radix.net, http://www.radix.net/~cknudsen/

Developers: go to the WebCalendar developer page on SourceForge:
  http://sourceforge.net/projects/webcalendar/

WebCalender = PHP-based multi-user calendar

See the INSTALL file for installation instructions.

See the FAQ file for frequently asked questions.

See UPGRADING for instructions on upgrading from a previous release.

See COPYING for licensing info.

See ChangeLog for a history of changes.

Features:
- Multi-user support
- View day-at-glance
- View month-at-glance
- View week-at-glance
- View year-at-glance
- View another user's calendar
- View one or more users' calendar via layers on top of your own calendar
- Add/Edit/Delete users
- Add/Edit/Delete events
- Repeating events
- Custom event fields
- Search interface for calendar entries
- User-configurable preferences for colors, 12/24 time format,
  Week start on Sun or Mon, default work hours
- Online help
- Checks for scheduling conflicts
- Users can accept or reject events added by another user to their calendar
- Email reminders
- Email notifications for new events
- Support for 17 languages: Chinese-Big5, Czech, Danish, Dutch,
  English, French, German, Icelandic, Italian, Korean, Norwegian, Polish,
  Portuguese, Portuguese/Brazil, Russian, Spanish, Turkish
- Exporting to Palm Pilot (via install-datebook, part of pilot-link)
  Note: install-datebook must be patched and rebuilt since it has
  a big old bug in it.  See INSTALL for instructions.
- Export to iCal

Requirements:
- PHP (3 or 4)
- Database: MySQL, Oracle, PostgreSQL or ODBC

Recommended:
- JavaScript-enabled browser
- CSS-enabled browser
- Cookie-enabled browser
- pilot-link (if exporting to Palm):
  http://sourceforge.net/project/?group_id=2158

Outstanding issues:
- No way to delete just a specific day of a repeating event
- No way to override the time of just a specific day of a repeating event
- Complete translatations (most are missing the translations for stuff
  related to repeating events)

Possible features for future releases:
- LDAP support
- User groups
- Custom access restrictions that can be defined by each user
- Some kind of shared public calendars.  Perhaps multiple shared public
  calendars that can be overlayed onto your own calendar.
- Allow users to not have JavaScript and/or CSS used
- Advances search (date ranges, and/or)
- Add comments to entries that will show up on the view page
- Data import (from Netscape Calendar's tab-delimted export or maybe
  from a Palm Pilot data file)
- Add support for repeating events to install-datebook
- Activity log
- Java client?


NEWS
WebCalendar NEWS File
Version: 1.2.8

This file lists significant changes to WebCalendar.  For a more detailed
list (that includes specific bugs fixed, please look at ChangeLog).

---------------------------------------------------------------------------
Version 1.2.9 (24 Aug 2017)
 - Fixed PHP compile error in login.php

Version 1.2.8 (11 Aug 2017)
 - Cross-site scripting (XSS) fixes.
 - Updates to supper PHP7.  PHP 5.3 is now minimum PHP version required.
 - Added missing code for mysqli support.
 - Gracefully handle if the function ob_gzhandler is not available.

Version 1.2.7 (22 Jan 2013)
 - Security fix: Do not show the reason for a failed login (i.e. "no such user")
 - Security fix: Escape HTML characters in category name.
 - Security fix: Check all passed in fields (either via HTML form or via
   URL parameter) for certain malicious tags (script, embed, etc.) and
   generate fatal error if found.

Version 1.2.6 (07 Jan 2013)
 - Fix for "undefined function require_valide" bug
 - Other minor bug fixes.
 - Updated Greek translation

Version 1.2.5 (29 Feb 2012)
 - Fixes for various security vulnerabilities include LFI (local file
   inclusion), XSS (cross site scripting) and others.
 - Updated translations: German, German-utf8, Dutch, Dutch-utf8
 - Misc. minor bug fixes

Version 1.2.4 (08 Aug 2011)
 - Fixed XSS vulnerability: malicious javascript in event descriptions submitted
   by public can do bad things (create admin account, delete events, etc.)
   when the pending event is viewed by the admin.
 - Fixed bug: PHP warnings on search
 - Removed PHP warnings
 - Bug fix: undefined function date_default_timezone_set in older versions
   of PHP.

Version 1.2.3 (14 Aug 2010)
 - Fixed bug: could not save events with more than one category
 - Include backtrace in fatal errors when in development mode

Version 1.2.2 (14 Aug 2010)
 - Fixes for PHP warnings about default timezone not being set.

Version 1.2.1 (09 Apr 2010)
 - Addressed some published security vulnerabilities related to
   cross site scripting (XSS).
 - Fixed "500 Server Error" seen with PHP 5.2.8 and later.  Some users saw
   either blank web page or seg fault errors with this.
 - Fixed mb_language error message some non-English users were seeing in 1.2.0.
 - A handful of minor bugfixes and patches.
 - Updated translations: Italian

Version 1.2.0 (27 Sep 2008)
 - Added new 'Security Audit' page.
 - Moved hard-coded upcoming.php config setting to System Settings.
 - Added new RSS feed for Activity Log.  This will allow you to be
   notified via RSS of event approvals/rejections, email notification,
   new event submissions, event updates, etc.
 - Removed old hack for register_globals work-around that is no longer
   needed.  This will improve security.
 - Category icons can now be displayed next to events in upcoming.php
 - Updated Portuguese_BR translations

Version 1.2.b1 (25 Feb 2008)
 - Changed default installation user/password to root/none
 - Added read/write test to db_cache /tmp folder
 - The following translations were updated:
   French, German-UTF

Version 1.1.6 (19 Sep 2007)

 - This release fixes bugs found in 1.1.5.
   (For a complete list of changes, see the ChangeLog in CVS.)
 - Security:
   Fixes cross-site scripting (XSS) vulnerability in export_handler.php.

Version 1.1.5 (12 Sep 2007)

  - Security fix: vulnerability in $includeDir in functions.php
  - The following translations were updated:
    Indonesian

Version 1.1.4 (28 Aug 2007)

  - This release fixes bugs found in 1.1.3.

Version 1.1.3 (04 Aug 2007)

  - New Features / Enhancements:
    = Import / Export / iCalendar / hCalendar / RSS
      + VTIMEZONE information is now captured from imported ics files and can be reused
        for future exports. This greatly simplifies the process and speeds up exports.
    = User Interface
      + Added option to display end times in calendars
      + Added CAPTCHA validation option  to Public Access event creation
  - The following translations were updated:
    German
    Basque

Version 1.1.2 (18 Dec 2006)

  - New Features / Enhancements:
    = Site Extras
      + Radio control are now available 
      + Checkbox controls are now available. 
      + Multi-Select Select controls are now available. 
      + Site-Extra fields are now included in emails. 
      + You can now set where each Site Extra fields show up. 
        - All 
        - Email 
        - View (view_entry) 
        - Reports 
        - Reminders 
        Any combination of the above is possible using bitwise operators.
    = User Interface
      + Date Range Searches have been added
      + Site Extra Seaches have been added
      + Category Filter has beed added to Advanced Search
      + Added Category Color option
      + Admin can now specify a background image for inclusion into CSS
      + Added new feature 'Display Long Day Names'
      + Added new Color Picker with many options
      + Added new About WebCalendar page
      + Reports can now contain 'fullname'
      + When creating Reports, you can now simply click desired value
        and it will auto-insert at the cursor.
      + The top menu is now fully configurable and custom menus can be
        added easily. See includes/menu/menu_config.php for details.
      + Added 'System Log' in menu.  Added logging for adding/editing/deleting
        users and invalid logins.
      + Reorganized and grouped System Settings and user preferences

  - The following translations were updated:
    French-UTF8(new), German, Russian

---------------------------------------------------------------------------

Version 1.1.1 (15 Sep 2006)

  - Fixes:
    + Fixed vulnerablity: SQL injection possible in edit_template.php,
      activity_log.php, admin_handler.php, pref_handler.php, export_handler.php
    + Fixed security: exploit allows execution of code downloaded from
      remote server
    + HUGE list of fixes (see ChangeLog for details)

  - New Features / Enhancements / Changes:

rss.php
<?php
/* $Id$
 *
 * Description:
 * This script is intended to be used outside of normal WebCalendar use,
 * as an RSS 2.0 feed to a RSS client.
 *
 * You must have "Enable RSS feed" set to "Yes"
 * in both System Settings and in the specific user's Preferences.
 *
 * Simply use the URL of this file as the feed address in the client.
 * For public user access:
 * http://xxxxx/aaa/rss.php
 * For any other user (where "joe" is the user login):
 * http:/xxxxxx/aaa/rss.php?user=joe
 *
 * By default (if you do not edit this file), events
 * will be loaded for either:
 *   - the next 30 days
 *   - the next 10 events
 *
 * Input parameters:
 * You can override settings by changing the URL parameters:
 *   - days: number of days ahead to look for events
 *   - cat_id: specify a category id to filter on
 *   - repeats: output all events including all repeat instances
 *       repeats=0 do not output repeating events (default)
 *       repeats=1 outputs repeating events
 *       repeats=2 outputs repeating events but suppresses display of
 *                 2nd & subsequent occurences of daily events
 *   - user: login name of calendar to display (instead of public user).
 *       You must have the following System Settings configured for this:
 *         Allow viewing other user's calendars: Yes
 *         Public access can view others: Yes
 *   - showdate: put the date and time (if specified in the title
 *       of the item) in the title
 *
 * Security:
 * $RSS_ENABLED must be set true
 * $USER_RSS_ENABLED must be set true unless this is for the public user
 * $USER_REMOTE_ACCESS can be set as follows in pref.php
 *      0 = Public entries only
 *      1 = Public & Confidential entries only
 *      2 = All entries are included in the feed *USE WITH CARE
 *
 * We do not include unapproved events in the RSS feed.
 *
 *
 * TODO
 * Add other RSS 2.0 options such as media.
 * Add <managingEditor>: dan@spam_me.com (Dan Deletekey)
 */

$debug = false;

include_once 'includes/translate.php';
require_once 'includes/classes/WebCalendar.class';
require_once 'includes/classes/Event.class';
require_once 'includes/classes/RptEvent.class';

$WebCalendar = new WebCalendar ( __FILE__ );

include_once 'includes/formvars.php';
include_once 'includes/functions.php';
include_once 'includes/config.php';
include_once 'includes/dbi4php.php';

$WebCalendar->initializeFirstPhase ();

include_once 'includes/' . $user_inc;

include_once 'includes/validate.php';
include_once 'includes/site_extras.php';

include_once 'includes/xcal.php';

$WebCalendar->initializeSecondPhase ();

load_global_settings ();

$WebCalendar->setLanguage ();

if ( empty ( $RSS_ENABLED ) || $RSS_ENABLED != 'Y' ) {
  header ( 'Content-Type: text/plain' );
  echo print_not_auth (28);
  exit;
}
/* Configurable settings for this file. You may change the settings below to
 * change the default settings. These settings will likely move into the System
 * Settings in the web admin interface in a future release.
 */

// Show the date in the title and how to format it.
$date_in_title = false; //Can override with "rss.php?showdate=1|true".
$showdate = getValue ( 'showdate' );
if ( ! empty ( $showdate ) )
  $date_in_title = ( $showdate == 'true' || $showdate == 1 ? true : false );

$date_format = 'M jS'; //Aug 10th, 8/10
$time_format = 'g:ia'; //4:30pm, 16:30
$time_separator = ', '; //Aug 10th @ 4:30pm, Aug 10th, 4:30pm

// Default time window of events to load.
// Can override with "rss.php?days=60".
$numDays = 30;

// Max number of events to display.
// Can override with "rss.php?max=20".
$maxEvents = 10;

// Login of calendar user to use.
// '__public__' is the login name for the public user.
$username = '__public__';

// Allow the URL to override the user setting such as "rss.php?user=craig".
$allow_user_override = true;

// Load layers.
$load_layers = false;

// Load just a specified category (by its id).
// Leave blank to not filter on category (unless specified in URL).
// Can override in URL with "rss.php?cat_id=4".
$cat_id = '';

// Load all repeating events.
// Can override with "rss.php?repeats=1".
$allow_repeats = false;

// Load show only first occurence within the given time span of daily repeating events.
// Can override with "rss.php?repeats=2".
$show_daily_events_only_once = false;

// End configurable settings...

// Set for use elsewhere as a global.
$login = $username;

if ( $allow_user_override ) {
  $u = getValue ( 'user', '[A-Za-z0-9_\.=@,\-]+', true );
  if ( ! empty ( $u ) ) {
    if ( $u == 'public' )
      $u = '__public__';
    // We also set $login since some functions assume that it is set..
    $login = $username = $u;
  }
}

load_user_preferences ();

// public entries only
$allow_access = array ( 'P' );

// .
// Determine what remote access has been set up by user.
// This will only be used if $username is not __public__.
if ( isset ( $USER_REMOTE_ACCESS ) && $username != '__public__' ) {
  if ( $USER_REMOTE_ACCESS > 0 ) // plus confidential
    $allow_access[] = 'C';

  if ( $USER_REMOTE_ACCESS == 2 ) // plus private
    $allow_access[] = 'R';
}
user_load_variables ( $login, 'rss_' );
$creator = ( $username == '__public__' ) ? 'Public' : $rss_fullname;

if ( $username != '__public__' &&
  ( empty ( $USER_RSS_ENABLED ) || $USER_RSS_ENABLED != 'Y' ) ) {
  header ( 'Content-Type: text/plain' );
  echo print_not_auth (29);
  exit;
}

$cat_id = '';
if ( $CATEGORIES_ENABLED == 'Y' ) {
  $x = getValue ( 'cat_id', '-?[0-9]+', true );
  if ( ! empty ( $x ) ) {
    load_user_categories ();
    $cat_id = $x;
    $category = $categories[$cat_id]['cat_name'];
  }
}

if ( $load_layers )
  load_user_layers ( $username );

// Calculate date range.
$date = getValue ( 'date', '-?[0-9]+', true );
if ( empty ( $date ) || strlen ( $date ) != 8 )
  // If no date specified, start with today.
  $date = date ( 'Ymd' );

$thisyear = substr ( $date, 0, 4 );
$thismonth = substr ( $date, 4, 2 );
$thisday = substr ( $date, 6, 2 );

$startTime = mktime ( 0, 0, 0, $thismonth, $thisday, $thisyear );

$x = getValue ( 'days', '-?[0-9]+', true );
if ( ! empty ( $x ) )
  $numDays = $x;

// Don't let a malicious user specify more than 365 days.
if ( $numDays > 365 )
  $numDays = 365;

$x = getValue ( 'max', '-?[0-9]+', true );
if ( ! empty ( $x ) )
  $maxEvents = $x;

// Don't let a malicious user specify more than 100 events.
if ( $maxEvents > 100 )
  $maxEvents = 100;

$x = getValue ( 'repeats', '-?[0-9]+', true );
if ( ! empty ( $x ) ) {
  $allow_repeats = $x;
  if ( $x == 2 )
    $show_daily_events_only_once = true;
}

$endTime = mktime ( 0, 0, 0, $thismonth, $thisday + $numDays -1, $thisyear );
$endDate = date ( 'Ymd', $endTime );

/* Pre-Load the repeated events for quicker access */
if ( $allow_repeats == true )
  $repeated_events = read_repeated_events ( $username, $startTime, $endTime, $cat_id );

/* Pre-load the non-repeating events for quicker access */
$events = read_events ( $username, $startTime, $endTime, $cat_id );

$charset = ( empty ( $LANGUAGE ) ? 'iso-8859-1' : translate ( 'charset' ) );
// This should work ok with RSS, may need to hardcode fallback value.
$lang = languageToAbbrev ( $LANGUAGE == 'Browser-defined' || $LANGUAGE == 'none'
  ? $lang : $LANGUAGE );
if ( $lang == 'en' )
  $lang = 'en-us'; //the RSS 2.0 default.

$appStr = generate_application_name ();

ob_start ();
// header ( 'Content-type: application/rss+xml');
header ( 'Content-type: text/xml' );
echo '<?xml version="1.0" encoding="' . $charset . '"?>
<rss version="2.0" xml:lang="' . $lang . '">
  <channel>
    <title><![CDATA[' . $appStr . ']]></title>
    <link>' . $SERVER_URL . '</link>
    <description><![CDATA[' . $appStr . ']]></description>
    <language>' . $lang . '</language>
    <generator>:"http://www.k5n.us/webcalendar.php?v=' . $PROGRAM_VERSION
 . '"</generator>
    <image>
      <title><![CDATA[' . $appStr . ']]></title>
      <link>' . $SERVER_URL . '</link>
      <url>http://www.k5n.us/k5n_small.gif</url>
    </image>';

$endtimeYmd = date ( 'Ymd', $endTime );
$numEvents = 0;
$reventIds = array ();
for ( $i = $startTime; date ( 'Ymd', $i ) <= $endtimeYmd && $numEvents < $maxEvents;
  $i += 86400 ) {
  $d = date ( 'Ymd', $i );
  $eventIds = array ();
  $pubDate = gmdate ( 'D, d M Y', $i );

  $entries = get_entries ( $d, false );
  $rentries = get_repeating_entries ( $username, $d );
  $entrycnt = count ( $entries );
  $rentrycnt = count ( $rentries );
  if ( $debug )
    echo '

countentries==' . $entrycnt . ' ' . $rentrycnt . '

';

  if ( $entrycnt > 0 || $rentrycnt > 0 ) {
    for ( $j = 0; $j < $entrycnt && $numEvents < $maxEvents; $j++ ) {
      // Prevent non-Public events from feeding
      if ( in_array ( $entries[$j]->getAccess (), $allow_access ) ) {
        $eventIds[] = $entries[$j]->getID ();
        $unixtime = $entries[$j]->getDateTimeTS ();
        $dateinfo = ( $date_in_title
          ? date ( $date_format, $unixtime )
           . ( $entries[$j]->isTimed () ? $time_separator 
		   . date ( $time_format, $unixtime ) : '' ) . ' ' : '' );

        echo '
    <item>
      <title><![CDATA[' . $dateinfo . $entries[$j]->getName () . ']]></title>
      <link>' . $SERVER_URL . 'view_entry.php?id=' . $entries[$j]->getID ()
         . '&amp;friendly=1&amp;rssuser=' . $login . '&amp;date=' . $d . '</link>
      <description><![CDATA[' . $entries[$j]->getDescription () . ']]></description>'
         . ( empty ( $category ) ? '' : '
      <category><![CDATA[' . $category . ']]></category>' )
        // . '<creator><![CDATA[' . $creator . ']]></creator>'
        /* RSS 2.0 date format Wed, 02 Oct 2002 13:00:00 GMT */. '
      <pubDate>' . gmdate ( 'D, d M Y H:i:s', $unixtime ) . ' GMT</pubDate>
      <guid>' . $SERVER_URL . 'view_entry.php?id=' . $entries[$j]->getID ()
         . '&amp;friendly=1&amp;rssuser=' . $login . '&amp;date=' . $d . '</guid>
    </item>';
        $numEvents++;
      }
    }
    for ( $j = 0; $j < $rentrycnt && $numEvents < $maxEvents; $j++ ) {
      // To allow repeated daily entries to be suppressed. Step below is
      // necessary because 1st occurence of repeating events shows up in
      // $entries AND $rentries & we suppress display of it in $rentries.
      if ( in_array ( $rentries[$j]->getID (),
            $eventIds ) && $rentries[$j]->getrepeatType () == 'daily' )
        $reventIds[] = $rentries[$j]->getID ();

      // Prevent non-Public events from feeding.
      // Prevent a repeating event from displaying if the original event has
      // already been displayed; prevent 2nd & later recurrence of daily events
      // from displaying if that option has been selected.
      if ( ! in_array ( $rentries[$j]->getID (), $eventIds ) &&
          ( ! $show_daily_events_only_once || ! in_array ( $rentries[$j]->getID (),
              $reventIds ) ) &&
          ( in_array ( $rentries[$j]->getAccess (), $allow_access ) ) ) {

        // Show repeating events only once.
        if ( $rentries[$j]->getrepeatType () == 'daily' )
          $reventIds[] = $rentries[$j]->getID ();

        echo '
    <item>';
        $unixtime = $rentries[$j]->getDateTimeTS ();
        // Constructing the TS for the current repeating event
        $unixtime = strtotime( date ( 'D, d M Y ', $i ) . date ( 'H:i:s', $unixtime ) ); 
        $dateinfo = ( $date_in_title
          ? date ( $date_format, $unixtime )
           . ( $rentries[$j]->isTimed ()
            ? $time_separator . date ( $time_format, $unixtime ) : '' ) . ' '
          : '' );

        echo '
      <title><![CDATA[' . $dateinfo . $rentries[$j]->getName () . ']]></title>
      <link>' . $SERVER_URL . "view_entry.php?id=" . $rentries[$j]->getID ()
         . '&amp;friendly=1&amp;rssuser=' . $login . '&amp;date=' . $d . '</link>
      <description><![CDATA[' . $rentries[$j]->getDescription () . ']]></description>'
         . ( empty ( $category ) ? '' : '
      <category><![CDATA[' . $category . ']]></category>' )
        // . '<creator><![CDATA[' . $creator . ']]></creator>'
        . '
      <pubDate>' . gmdate ( 'D, d M Y H:i:s', $unixtime ) . ' GMT</pubDate>
      <guid>' . $SERVER_URL . 'view_entry.php?id=' . $rentries[$j]->getID ()
         . '&amp;friendly=1&amp;rssuser=' . $login . '&amp;date=' . $d . '</guid>
    </item>';
        $numEvents++;
      }
    }
  }
}
echo '
  </channel>
</rss>';
ob_end_flush ();
// Clear login...just in case.
$login = '';
exit;

?>
